joomla/joomla-cms
# Action logs CSV export with Generator support

Pull Request for Issue #22700

### Summary of Changes

This builds on #22750 to add a separate code path for PHP 5.5+ users to use the more memory efficient Generators for the CSV data.  As we are building a data set that is iterated over one time only, this fits the use case for a Generator, unfortunately it is a feature newer than our minimum requirements.

Generally, we would not accept a solution that creates different code paths for different PHP versions, but I think we're reaching a point where it's becoming OK to create special exemptions for special circumstances, this being one of them.

- Of sites reporting stats and running 3.8.13, 2.2% of those sites are using PHP 5.3 or 5.4 (and 5.2% of all sites having sent an update in the last 90 days across all versions), therefore this code path is for the benefit of an overwhelming majority of users
- While this creates a differing return type depending on the PHP version, the intended use case for it (doing a one-time `foreach` loop over the result) is not impacted

See #22750 for info on the first commit of this PR.

### Testing Instructions

Create a large database table of action logs to export, try to export across PHP versions.  It should work for all versions and on 5.5 and newer use the memory optimized code path.  Using the database @PhilETaylor supplied in #22700 I cannot get the export to complete on a PHP 7.2 system with 256M memory limit, using a Generator I can.
